src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,121 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] items;
+    private int size;
+
+    public ArrayList() {
+        this.items = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
+        if (size == items.length) {
+            items = increaseCapacity();
+        }
 
+        items[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+
+        if (size == items.length) {
+            items = increaseCapacity();
+        }
+
+        System.arraycopy(items, index, items, index + 1, size - index);
+
+        items[index] = value;
 
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+
+        return getElementByIndex(index);
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        items[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+
+        T oldValue = getElementByIndex(index);
+
+        int numMoved = size - index - 1;
+
+        System.arraycopy(items, index + 1, items, index, numMoved);
+
+        size--;
+
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
```

Comment:
The `remove(T element)` method should explicitly handle the case where `element` is null. Currently, it relies on the `equals` method, which may not be appropriate for all types.